scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala

```diff
@@ -494,15 +532,61 @@ object KRouter {
       }
   }
 
-  // These node records are derived from Ethereum node records (https://eips.ethereum.org/EIPS/eip-778)
-  // TODO node records require an additional
-  // signature (why)
-  // sequence number (why)
-  // compressed public key (why)
-  // TODO understand what these things do, which we need an implement.
-  case class NodeRecord[A](id: BitVector, routingAddress: A, messagingAddress: A) {
+  /**
+    * These node records are derived from Ethereum node records (https://eips.ethereum.org/EIPS/eip-778)
+    * @param id Kademlia routing direction (compressed public key).
+    * @param seq instance number of the NodeRecord.
+    * @param sign signature of the rest of NodeRecord
+    */
+  case class NodeRecord[A](id: BitVector, routingAddress: A, messagingAddress: A, seq: Long, sign: ECDSASignature) {
     override def toString: String =
-      s"NodeRecord(id = ${id.toHex}, routingAddress = $routingAddress, messagingAddress = $messagingAddress)"
+      s"NodeRecord(sec_number = ${seq}, id = ${id.toHex}, routingAddress = $routingAddress, messagingAddress = $messagingAddress)"
+  }
+
+  object NodeRecord {
+    def apply[A](
+        id: BitVector,
+        routingAddress: A,
+        messagingAddress: A,
+        sec_number: Long,
+        keyPair: AsymmetricCipherKeyPair
+    )(implicit codec: Codec[A]): NodeRecord[A] = {
+      val encodedUUID = ByteBuffer.allocate(8)
+      encodedUUID.putLong(0, sec_number)
+      val encoded = id.toByteArray ++ codec
+        .encode(routingAddress)
+        .toOption
+        .get
+        .toByteArray ++ codec
+        .encode(routingAddress)
+        .toOption
+        .get
+        .toByteArray ++ encodedUUID.array()
+      val sign = crypto.ECDSASignature.sign(encoded, keyPair)
+      NodeRecord[A](id, routingAddress, messagingAddress, sec_number, sign)
+    }
+    def verify[A](n: NodeRecord[A])(implicit codec: Codec[A]): Boolean = {
+      val encodedUUID = ByteBuffer.allocate(8)
+      encodedUUID.putLong(0, n.seq)
+      val encoded = n.id.toByteArray ++ codec
```

Comment:
seems like a repetion from `apply` above, also the intet is not clear to me here. Maybe it would be better to move it to separate function ? (also i would try to avoid using direct bytebuffer calls)